pkg/controller/conversion/functions.go

```diff
@@ -17,24 +17,27 @@ import (
 // representation of the `src` object and applies the registered webhook
 // conversion functions of this registry.
 func (r *registry) RoundTrip(dst, src resource.Terraformed) error { //nolint:gocyclo // considered breaking this according to the converters and I did not like it
+	// first PrioritizedManagedConversions are run in their registration order
+	for _, c := range r.GetConversions(dst) {
+		if pc, ok := c.(conversion.PrioritizedManagedConversion); ok {
+			if _, err := pc.ConvertManaged(src, dst); err != nil {
+				return errors.Wrapf(err, "cannot apply the PrioritizedManagedConversion for the %q object", dst.GetTerraformResourceType())
+			}
+		}
+	}
+
 	srcMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(src)
 	if err != nil {
 		return errors.Wrap(err, "cannot convert the conversion source object into the map[string]any representation")
 	}
-	gvk := dst.GetObjectKind().GroupVersionKind()
-	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(srcMap, dst); err != nil {
-		return errors.Wrap(err, "cannot convert the map[string]any representation of the source object to the conversion target object")
-	}
-	// restore the original GVK for the conversion destination
-	dst.GetObjectKind().SetGroupVersionKind(gvk)
-
```

Comment:
Yes, correct. Previously, `conversion.RoundTrip` would unconditionally do the identity conversion which copies fields with the same name & path into the target object. But in the scenario where we have converted singleton lists into embedded objects, the source and the target end up having fields with different types (a list and an object) at identical paths, which breaks the identity conversion. 

Furthermore, it's possible to have, for instance, a `conversion.ManagedConversion`, which performs the identity conversion itself (e.g., by directly copying some `metadata` fields from source into the target). We previously did not have this flexibility and `conversion.RoundTrip` would always do the identity conversion.

It's now optional and for backwards compatibility, `config.DefaultResource` now initializes the `config.Resource.Conversions` slice with the identity conversion. In contexts where we don't want the identity conversion, then we need to replace `Conversions` slice. 

There could in theory be clients which just replace the slice here although they need the identity conversion and will be broken when they consume these changes. If they are overriding the default `Conversions` slice, then it means they are not interested in using the defaults. But probably they just assume there are no defaults instead of not _using_ these (or any future) defaults. I consider this as a bad practice. Another complication is that there could be multiple places in the provider code where the converters are added. If multiple sites replace these converters (instead of appending to already existing converters), then again it's probably unintended.

Unfortunately, I think this has roots in not following the principle of [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)) in the resource configuration API (the configuration data fields are just exported). This requires a self-discipline at the client side.